lib/src/gitattributes.rs

```diff
@@ -0,0 +1,705 @@
+        disk_file_loader: impl FileLoader + 'static,
+    ) -> Self {
+        Self {
+            store_file_loader: Arc::new(store_file_loader),
+            disk_file_loader: Arc::new(disk_file_loader),
+            node_cache: Default::default(),
+        }
+    }
+
+    pub(crate) async fn search(
+        &self,
+        path: &RepoPath,
+        attribute_names: impl AsRef<[&str]>,
+        priority: SearchPriority,
+    ) -> Result<HashMap<String, State>, GitAttributesError> {
+        let attributes = self.get_git_attributes_node(path);
+        let store = match priority {
+            SearchPriority::Store => attributes.get_store_first().await?,
+            SearchPriority::Disk => attributes.get_disk_first().await?,
+        };
+        let (search, collection) = &*store;
+
+        let mut out = Outcome::default();
+        out.initialize_with_selection(
+            collection,
+            // From<&str> is implemented for KStringRef
+            attribute_names.as_ref().iter().copied(),
+        );
+        search.pattern_matching_relative_path(
+            path.as_internal_file_string().into(),
+            Case::Sensitive,
+            None,
+            &mut out,
+        );
+
+        let mut map = out
+            .iter_selected()
+            .map(|attr| {
+                let val = attr.assignment.to_owned();
+                (val.name.as_str().to_string(), val.state)
+            })
+            .collect::<HashMap<String, State>>();
+
+        // go over attributes and mark as unspecified if not set
+        for attribute_name in attribute_names.as_ref() {
+            map.entry(attribute_name.to_string())
+                .or_insert(State::Unspecified);
+        }
+
+        Ok(map)
+    }
+
+    fn get_git_attributes_node(&self, path: &RepoPath) -> Arc<GitAttributesNode> {
+        let mut val = self.node_cache.lock().expect("Not be poisoned");
+        self.inner(&mut val, path)
+    }
+
+    fn inner(
+        &self,
+        map: &mut HashMap<RepoPathBuf, Arc<GitAttributesNode>>,
+        path: &RepoPath,
+    ) -> Arc<GitAttributesNode> {
+        if let Some(node) = map.get(path) {
+            return node.clone();
+        }
+        let parent = path.parent().map(|parent| self.inner(map, parent));
+        let new_node = Arc::new(GitAttributesNode {
+            disk_first: OnceCell::new(),
+            store_first: OnceCell::new(),
+            disk_file_loader: self.disk_file_loader.clone(),
+            store_file_loader: self.store_file_loader.clone(),
+            parent,
+            path: path.to_owned(),
+        });
+        map.insert(path.to_owned(), new_node.clone());
+        new_node
+    }
+}
+
+/// Errors for GitAttributes
+#[derive(Debug, thiserror::Error)]
+#[error("{message}")]
+pub struct GitAttributesError {
+    message: String,
+    #[source]
+    source: Box<dyn std::error::Error + Send + Sync>,
+}
+
+type TestStore = HashMap<RepoPathBuf, Result<String, String>>;
```

Comment:
Move `type TestStore` and `impl FileLoader for TestStore` to `mod tests`. Rust should allow such structure. And I assume that we only need such store in unit tests.